### ISSUE TITLE
"No space left on device" despite moveonenospc=true

### ISSUE THREAD

**Issue by broken** [trapexit/mergerfs#1327]
**Describe the bug**

The setting `moveonenospc=true` does not automatically create directories on a new disk when a current one gets filled. Changing the create policy to mfs does create the file as expected on the empty drive.

**To Reproduce**

Settings:
/mnt/disk* /mnt/storage fuse.mergerfs defaults,moveonenospc=true,minfreespace=50G,fsname=mergerfs 0 0

Steps:
cd to a path that exists on a full drive and not a drive with space.
```
$ touch test
touch: cannot touch 'test': No space left on device
```

**Expected behavior**

I would expect that the test file and paths are created on the drive with space.
I tried `moveonenospc=mfs` as well.

If I set `category.create=mfs` the file is created on the drive with available space as expected.

I suspect this is either a case of:
 (1) Despite reading the documentation, I am not understanding the option correctly.
 (2) A problem with my filesystem. IIUC, mergerfs needs to receive a ENOSPC error to enact the mfs policy. I am not seeing this from the strace, so mergerfs never attempts to find a new drive.

**System information:**

 - OS, kernel version: `uname -a`
 `Linux vision 5.10.0-27-amd64 #1 SMP Debian 5.10.205-2 (2023-12-31) x86_64 GNU/Linux`

 - mergerfs version: `mergerfs -V`
`mergerfs v2.40.2`

 - mergerfs settings
`/mnt/disk* /mnt/storage fuse.mergerfs defaults,moveonenospc=true,minfreespace=50G,fsname=mergerfs 0 0`

 - List of drives, filesystems, & sizes:
   - `df -h`
```
Filesystem                   Size  Used Avail Use% Mounted on
udev                         7.8G     0  7.8G   0% /dev
tmpfs                        1.6G  3.0M  1.6G   1% /run
/dev/mapper/vision--vg-root  442G   27G  393G   7% /
tmpfs                        7.8G     0  7.8G   0% /dev/shm
tmpfs                        5.0M     0  5.0M   0% /run/lock
mergerfs                      40T   33T  7.0T  83% /mnt/storage
/dev/sdf1                    7.3T   32K  6.9T   1% /mnt/disk7
/dev/sdi1                    7.3T  7.3T     0 100% /mnt/disk4
/dev/sdh1                    7.3T  7.2T   26G 100% /mnt/disk5
/dev/sdg1                    7.3T  7.3T     0 100% /mnt/disk6
/dev/sdb2                    235M   89M  134M  40% /boot
/dev/sde1                    7.3T  2.3T  5.0T  32% /mnt/parity1
/dev/sda1                    3.6T  3.6T     0 100% /mnt/disk1
/dev/sdc1                    3.6T  3.6T   35G 100% /mnt/disk2
/dev/sdd1                    3.6T  3.6T     0 100% /mnt/disk3
/dev/sdb1                    511M  5.8M  506M   2% /boot/efi
tmpfs                        1.6G     0  1.6G   0% /run/user/1000
```

   - `lsblk -f`
```
NAME      FSTYPE      FSVER    LABEL UUID                                   FSAVAIL FSUSE% MOUNTPOINT
sda
└─sda1    ext4        1.0            d227df72-5f53-467d-8c0c-81472ad6b8ad         0   100% /mnt/disk1
sdb
├─sdb1    vfat        FAT32          B167-E095                               505.2M     1% /boot/efi
├─sdb2    ext2        1.0            924f8461-4434-4d04-af0e-305479118723    133.5M    38% /boot
└─sdb3    LVM2_member LVM2 001       zKhtyf-zeas-JoqB-Y68w-R5cI-N0p6-WTFav7
  ├─vision--vg-root
  │       ext4        1.0            6f4e5d6e-976e-402a-9766-6d846c5cc2c2    392.5G     6% /
  └─vision--vg-swap_1
          swap        1              d7b50457-0e4d-44aa-9cc9-8e6c2583314b                  [SWAP]
sdc
└─sdc1    ext4        1.0            fab0ee01-b018-4352-ae9e-19325ea6cae6     34.7G    99% /mnt/disk2
sdd
└─sdd1    ext4        1.0            6116db89-82df-4366-b4e2-5f1cc3d189a6         0   100% /mnt/disk3
sde
└─sde1    ext4        1.0            41775f45-1712-46df-97b5-a416269c914a      4.9T    32% /mnt/parity1
sdf
└─sdf1    ext4        1.0            6615e493-c8d1-4bcd-a37b-90e55f167e93      6.9T     0% /mnt/disk7
sdg
└─sdg1    ext4        1.0            289ef181-e17f-48bb-912b-aa1b84452b61         0   100% /mnt/disk6
sdh
└─sdh1    ext4        1.0            d3900d5b-d22e-448f-878e-16a9915eb030     25.5G   100% /mnt/disk5
sdi
└─sdi1    ext4        1.0            8ab8e65b-3d29-4dfe-b870-3f4c490718a0         0   100% /mnt/disk4
```

 - A strace of the application having a problem:
   - `strace -fvTtt -s 256 -o /tmp/app.strace.txt touch test`
```
2899  21:46:33 execve("/usr/bin/touch", ["touch", "test"], ["SHELL=/bin/bash", "PUID=1001", "PWD=/mnt/storage/tmp", "LOGNAME=dogatech", "XDG_SESSION_TYPE=tty", "MOTD_SHOWN=pam", "HOME=/root", "LANG=en_US.UTF-8", "LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*"..., "PGID=1001", "LC_TERMINAL=iTerm2", "SSH_CONNECTION=192.168.1.34 55315 192.168.1.2 22", "XDG_SESSION_CLASS=user", "TERM=xterm-256color", "USER=dogatech", "LC_TERMINAL_VERSION=3.4.23", "SHLVL=2", "XDG_SESSION_ID=3", "XDG_RUNTIME_DIR=/run/user/1000", "SSH_CLIENT=192.168.1.34 55315 22", "PATH=/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/usr/local/sbin:/usr/sbin:/sbin", "DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus", "MAIL=/var/mail/root", "SSH_TTY=/dev/pts/1", "OLDPWD=/mnt/disk7", "_=/usr/bin/strace"]) = 0 <0.000620>
2899  21:46:33 brk(NULL)                = 0x55ed7f5c9000 <0.000128>
2899  21:46:33 access("/etc/ld.so.preload", R_OK) = -1 ENOENT (No such file or directory) <0.000137>
2899  21:46:33 openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3 <0.000105>
2899  21:46:33 fstat(3, {st_dev=makedev(0xfe, 0), st_ino=8126508, st_mode=S_IFREG|0644, st_nlink=1, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=88, st_size=42735, st_atime=1712914880 /* 2024-04-12T02:41:20.973117553-0700 */, st_atime_nsec=973117553, st_mtime=1706607294 /* 2024-01-30T01:34:54.866299533-0800 */, st_mtime_nsec=866299533, st_ctime=1706607295 /* 2024-01-30T01:34:55.086298979-0800 */, st_ctime_nsec=86298979}) = 0 <0.000082>
2899  21:46:33 mmap(NULL, 42735, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f009327e000 <0.000087>
2899  21:46:33 close(3)                 = 0 <0.000075>
2899  21:46:33 openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3 <0.000044>
2899  21:46:33 read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0@>\2\0\0\0\0\0@\0\0\0\0\0\0\0\340\363\34\0\0\0\0\0\0\0\0\0@\08\0\f\0@\0@\0?\0\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0\240\2\0\0\0\0\0\0\240\2\0\0\0\0\0\0\10\0\0\0\0\0\0\0\3\0\0\0\4\0\0\0P\366\31\0\0\0\0\0P\366\31\0\0\0\0\0P\366\31\0\0\0\0\0\34\0\0\0\0\0\0\0\34\0\0\0\0\0\0\0\20\0\0\0\0\0\0\0\1\0\0\0\4\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\210\24\2\0\0\0\0\0\210\24\2\0\0\0\0\0\0\20\0\0\0\0\0\0\1\0\0\0\5\0\0\0\0 \2\0\0\0\0\0\0 \2\0\0\0\0\0"..., 832) = 832 <0.000027>
2899  21:46:33 fstat(3, {st_dev=makedev(0xfe, 0), st_ino=6424217, st_mode=S_IFREG|0755, st_nlink=1, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=3720, st_size=1901536, st_atime=1712897507 /* 2024-04-11T21:51:47.317127688-0700 */, st_atime_nsec=317127688, st_mtime=1696278177 /* 2023-10-02T13:22:57-0700 */, st_mtime_nsec=0, st_ctime=1699227502 /* 2023-11-05T15:38:22.653190891-0800 */, st_ctime_nsec=653190891}) = 0 <0.000026>
2899  21:46:33 mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f009327c000 <0.000031>
2899  21:46:33 mmap(NULL, 1914496, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f00930a8000 <0.000031>
2899  21:46:33 mmap(0x7f00930ca000, 1413120, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x22000) = 0x7f00930ca000 <0.000047>
2899  21:46:33 mmap(0x7f0093223000, 323584, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x17b000) = 0x7f0093223000 <0.000035>
2899  21:46:33 mmap(0x7f0093272000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1c9000) = 0x7f0093272000 <0.000035>
2899  21:46:33 mmap(0x7f0093278000, 13952, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f0093278000 <0.000031>
2899  21:46:33 close(3)                 = 0 <0.000024>
2899  21:46:33 mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f00930a6000 <0.000028>
2899  21:46:33 arch_prctl(ARCH_SET_FS, 0x7f009327d5c0) = 0 <0.000023>
2899  21:46:33 mprotect(0x7f0093272000, 16384, PROT_READ) = 0 <0.000045>
2899  21:46:33 mprotect(0x55ed7d61a000, 4096, PROT_READ) = 0 <0.000032>
2899  21:46:33 mprotect(0x7f00932b3000, 4096, PROT_READ) = 0 <0.000032>
2899  21:46:33 munmap(0x7f009327e000, 42735) = 0 <0.000053>
2899  21:46:33 brk(NULL)                = 0x55ed7f5c9000 <0.000024>
2899  21:46:33 brk(0x55ed7f5ea000)      = 0x55ed7f5ea000 <0.000028>
2899  21:46:33 openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3 <0.000035>
2899  21:46:33 fstat(3, {st_dev=makedev(0xfe, 0), st_ino=6429363, st_mode=S_IFREG|0644, st_nlink=1, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=5800, st_size=3041456, st_atime=1712926801 /* 2024-04-12T06:00:01.382771084-0700 */, st_atime_nsec=382771084, st_mtime=1699227646 /* 2023-11-05T15:40:46.491972236-0800 */, st_mtime_nsec=491972236, st_ctime=1699227646 /* 2023-11-05T15:40:46.491972236-0800 */, st_ctime_nsec=491972236}) = 0 <0.000025>
2899  21:46:33 mmap(NULL, 3041456, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f0092dbf000 <0.000030>
2899  21:46:33 close(3)                 = 0 <0.000023>
2899  21:46:33 openat(AT_FDCWD, "test", O_WRONLY|O_CREAT|O_NOCTTY|O_NONBLOCK, 0666) = -1 ENOSPC (No space left on device) <0.000398>
2899  21:46:33 utimensat(AT_FDCWD, "test", NULL, 0) = -1 ENOENT (No such file or directory) <0.000174>
2899  21:46:33 openat(AT_FDCWD, "/usr/share/locale/locale.alias", O_RDONLY|O_CLOEXEC) = 3 <0.000041>
2899  21:46:33 fstat(3, {st_dev=makedev(0xfe, 0), st_ino=8127489, st_mode=S_IFREG|0644, st_nlink=1, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=8, st_size=2996, st_atime=1712934063 /* 2024-04-12T08:01:03.361032269-0700 */, st_atime_nsec=361032269, st_mtime=1661549566 /* 2022-08-26T14:32:46-0700 */, st_mtime_nsec=0, st_ctime=1663012597 /* 2022-09-12T12:56:37.439065616-0700 */, st_ctime_nsec=439065616}) = 0 <0.000026>
2899  21:46:33 read(3, "# Locale name alias data base.\n# Copyright (C) 1996-2020 Free Software Foundation, Inc.\n#\n# This program is free software; you can redistribute it and/or modify\n# it under the terms of the GNU General Public License as published by\n# the Free Software Foun"..., 4096) = 2996 <0.000028>
2899  21:46:33 read(3, "", 4096)        = 0 <0.000025>
2899  21:46:33 close(3)                 = 0 <0.000026>
2899  21:46:33 openat(AT_FDCWD, "/usr/share/locale/en_US.UTF-8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000031>
2899  21:46:33 openat(AT_FDCWD, "/usr/share/locale/en_US.utf8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000028>
2899  21:46:33 openat(AT_FDCWD, "/usr/share/locale/en_US/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000030>
2899  21:46:33 openat(AT_FDCWD, "/usr/share/locale/en.UTF-8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000028>
2899  21:46:33 openat(AT_FDCWD, "/usr/share/locale/en.utf8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000028>
2899  21:46:33 openat(AT_FDCWD, "/usr/share/locale/en/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000030>
2899  21:46:33 write(2, "touch: ", 7)   = 7 <0.000057>
2899  21:46:33 write(2, "cannot touch 'test'", 19) = 19 <0.000052>
2899  21:46:33 openat(AT_FDCWD, "/usr/share/locale/en_US.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000043>
2899  21:46:33 openat(AT_FDCWD, "/usr/share/locale/en_US.utf8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000027>
2899  21:46:33 openat(AT_FDCWD, "/usr/share/locale/en_US/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000029>
2899  21:46:33 openat(AT_FDCWD, "/usr/share/locale/en.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000027>
2899  21:46:33 openat(AT_FDCWD, "/usr/share/locale/en.utf8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000027>
2899  21:46:33 openat(AT_FDCWD, "/usr/share/locale/en/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory) <0.000028>
2899  21:46:33 write(2, ": No space left on device", 25) = 25 <0.000038>
2899  21:46:33 write(2, "\n", 1)        = 1 <0.000044>
2899  21:46:33 close(1)                 = 0 <0.000026>
2899  21:46:33 close(2)                 = 0 <0.000024>
2899  21:46:33 exit_group(1)            = ?
2899  21:46:33 +++ exited with 1 +++
```

 - strace of mergerfs while app tried to do it's thing:
   - `strace -fvTtt -s 256 -p <mergerfsPID> -o /tmp/mergerfs.strace.txt`
```
330   21:16:23.082369 read(4,  <unfinished ...>
329   21:16:23.082533 read(4,  <unfinished ...>
328   21:16:23.082575 read(4,  <unfinished ...>
326   21:16:23.082618 restart_syscall(<... resuming interrupted read ...> <unfinished ...>
327   21:16:23.082647 read(4,  <unfinished ...>
325   21:16:23.082689 futex(0x7ffdbe5934e0, FUTEX_WAIT_BITSET_PRIVATE|FUTEX_CLOCK_REALTIME, 0, NULL, FUTEX_BITSET_MATCH_ANY <unfinished ...>
330   21:16:26.822895 <... read resumed>"8\0\0\0\3\0\0\0004\1\0\0\0\0\0\0\2\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\30\v\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 1052672) = 56 <3.740364>
330   21:16:26.823073 lstat("/mnt/disk1/tmp", 0x7f360e7f3200) = -1 ENOENT (No such file or directory) <0.000051>
330   21:16:26.823471 lstat("/mnt/disk2/tmp", 0x7f360e7f3200) = -1 ENOENT (No such file or directory) <0.000047>
330   21:16:26.823642 lstat("/mnt/disk3/tmp", 0x7f360e7f3200) = -1 ENOENT (No such file or directory) <0.000046>
330   21:16:26.823787 lstat("/mnt/disk4/tmp", {st_dev=makedev(0x8, 0x81), st_ino=80805889, st_mode=S_IFDIR|0755, st_nlink=7, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=8, st_size=4096, st_atime=1712979967 /* 2024-04-12T20:46:07.044004208-0700 */, st_atime_nsec=44004208, st_mtime=1531192660 /* 2018-07-09T20:17:40.633277693-0700 */, st_mtime_nsec=633277693, st_ctime=1531192660 /* 2018-07-09T20:17:40.633277693-0700 */, st_ctime_nsec=633277693}) = 0 <0.000037>
330   21:16:26.823997 lstat("/mnt/disk4/tmp", {st_dev=makedev(0x8, 0x81), st_ino=80805889, st_mode=S_IFDIR|0755, st_nlink=7, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=8, st_size=4096, st_atime=1712979967 /* 2024-04-12T20:46:07.044004208-0700 */, st_atime_nsec=44004208, st_mtime=1531192660 /* 2018-07-09T20:17:40.633277693-0700 */, st_mtime_nsec=633277693, st_ctime=1531192660 /* 2018-07-09T20:17:40.633277693-0700 */, st_ctime_nsec=633277693}) = 0 <0.000033>
330   21:16:26.824192 writev(4, [{iov_base="x\0\0\0\0\0\0\0004\1\0\0\0\0\0\0", iov_len=16}, {iov_base="\1\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\231\246\"\325\230\376\224\\\0\20\0\0\0\0\0\0\10\0\0\0\0\0\0\0\377\377\31f\0\0\0\0T%D[\0\0\0\0T%D[\0\0\0\0ps\237\2\375\f\277%\375\f\277%\355A\0\0\7\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\20\0\0\0\0\0\0", iov_len=104}], 2) = 120 <0.000041>
330   21:16:26.824340 read(4,  <unfinished ...>
329   21:16:26.824396 <... read resumed>"-\0\0\0\1\0\0\0006\1\0\0\0\0\0\0\2\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\30\v\0\0\0\0\0\0test\0", 1052672) = 45 <3.741828>
329   21:16:26.824486 lstat("/mnt/disk1/tmp/test", 0x7f360eff41f0) = -1 ENOENT (No such file or directory) <0.000033>
329   21:16:26.824619 lstat("/mnt/disk2/tmp/test", 0x7f360eff41f0) = -1 ENOENT (No such file or directory) <0.000031>
329   21:16:26.824746 lstat("/mnt/disk3/tmp/test", 0x7f360eff41f0) = -1 ENOENT (No such file or directory) <0.000031>
329   21:16:26.824872 lstat("/mnt/disk4/tmp/test", 0x7f360eff41f0) = -1 ENOENT (No such file or directory) <0.000032>
329   21:16:26.824998 lstat("/mnt/disk5/tmp/test", 0x7f360eff41f0) = -1 ENOENT (No such file or directory) <0.000031>
329   21:16:26.825122 lstat("/mnt/disk6/tmp/test", 0x7f360eff41f0) = -1 ENOENT (No such file or directory) <0.000032>
329   21:16:26.825249 lstat("/mnt/disk7/tmp/test", 0x7f360eff41f0) = -1 ENOENT (No such file or directory) <0.000031>
329   21:16:26.825375 writev(4, [{iov_base="\220\0\0\0\0\0\0\0006\1\0\0\0\0\0\0", iov_len=16}, {iov_base="\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\1\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", iov_len=128}], 2) = 144 <0.000112>
329   21:16:26.825632 read(4,  <unfinished ...>
328   21:16:26.825677 <... read resumed>"=\0\0\0#\0\0\08\1\0\0\0\0\0\0\2\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\30\v\0\0\0\0\0\0A\210\0\0\266\201\0\0\22\0\0\0\0\0\0\0test\0", 1052672) = 61 <3.743064>
328   21:16:26.825787 lstat("/mnt/disk1/tmp", 0x7f360f7f5180) = -1 ENOENT (No such file or directory) <0.000085>
328   21:16:26.826026 lstat("/mnt/disk2/tmp", 0x7f360f7f5180) = -1 ENOENT (No such file or directory) <0.000084>
328   21:16:26.826231 lstat("/mnt/disk3/tmp", 0x7f360f7f5180) = -1 ENOENT (No such file or directory) <0.000028>
328   21:16:26.826351 lstat("/mnt/disk4/tmp", {st_dev=makedev(0x8, 0x81), st_ino=80805889, st_mode=S_IFDIR|0755, st_nlink=7, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=8, st_size=4096, st_atime=1712979967 /* 2024-04-12T20:46:07.044004208-0700 */, st_atime_nsec=44004208, st_mtime=1531192660 /* 2018-07-09T20:17:40.633277693-0700 */, st_mtime_nsec=633277693, st_ctime=1531192660 /* 2018-07-09T20:17:40.633277693-0700 */, st_ctime_nsec=633277693}) = 0 <0.000033>
328   21:16:26.826543 lstat("/mnt/disk1/tmp", 0x7f360f7f5180) = -1 ENOENT (No such file or directory) <0.000027>
328   21:16:26.826658 lstat("/mnt/disk2/tmp", 0x7f360f7f5180) = -1 ENOENT (No such file or directory) <0.000027>
328   21:16:26.826773 lstat("/mnt/disk3/tmp", 0x7f360f7f5180) = -1 ENOENT (No such file or directory) <0.000027>
328   21:16:26.826888 lstat("/mnt/disk4/tmp", {st_dev=makedev(0x8, 0x81), st_ino=80805889, st_mode=S_IFDIR|0755, st_nlink=7, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=8, st_size=4096, st_atime=1712979967 /* 2024-04-12T20:46:07.044004208-0700 */, st_atime_nsec=44004208, st_mtime=1531192660 /* 2018-07-09T20:17:40.633277693-0700 */, st_mtime_nsec=633277693, st_ctime=1531192660 /* 2018-07-09T20:17:40.633277693-0700 */, st_ctime_nsec=633277693}) = 0 <0.000028>
328   21:16:26.827065 statfs("/mnt/disk4", {f_type=EXT2_SUPER_MAGIC, f_bsize=4096, f_blocks=1937818044, f_bfree=3909, f_bavail=0, f_files=244191232, f_ffree=244154233, f_fsid={val=[400148530, 1582640756]}, f_namelen=255, f_frsize=4096, f_flags=ST_VALID|ST_RELATIME}) = 0 <0.000030>
328   21:16:26.827226 lstat("/mnt/disk5/tmp", 0x7f360f7f5180) = -1 ENOENT (No such file or directory) <0.000028>
328   21:16:26.827351 lstat("/mnt/disk6/tmp", 0x7f360f7f5180) = -1 ENOENT (No such file or directory) <0.000027>
328   21:16:26.827469 lstat("/mnt/disk7/tmp", 0x7f360f7f5180) = -1 ENOENT (No such file or directory) <0.000027>
328   21:16:26.827587 writev(4, [{iov_base="\20\0\0\0\344\377\377\3778\1\0\0\0\0\0\0", iov_len=16}], 1) = 16 <0.000042>
328   21:16:26.827728 read(4,  <unfinished ...>
327   21:16:26.827774 <... read resumed>"-\0\0\0\1\0\0\0:\1\0\0\0\0\0\0\2\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\30\v\0\0\0\0\0\0test\0", 1052672) = 45 <3.745104>
327   21:16:26.827866 lstat("/mnt/disk1/tmp/test", 0x7f360fff61f0) = -1 ENOENT (No such file or directory) <0.000037>
327   21:16:26.828000 lstat("/mnt/disk2/tmp/test", 0x7f360fff61f0) = -1 ENOENT (No such file or directory) <0.000032>
327   21:16:26.828123 lstat("/mnt/disk3/tmp/test", 0x7f360fff61f0) = -1 ENOENT (No such file or directory) <0.000027>
327   21:16:26.828239 lstat("/mnt/disk4/tmp/test", 0x7f360fff61f0) = -1 ENOENT (No such file or directory) <0.000028>
327   21:16:26.828355 lstat("/mnt/disk5/tmp/test", 0x7f360fff61f0) = -1 ENOENT (No such file or directory) <0.000027>
327   21:16:26.828470 lstat("/mnt/disk6/tmp/test", 0x7f360fff61f0) = -1 ENOENT (No such file or directory) <0.000026>
327   21:16:26.828586 lstat("/mnt/disk7/tmp/test", 0x7f360fff61f0) = -1 ENOENT (No such file or directory) <0.000027>
327   21:16:26.828704 writev(4, [{iov_base="\220\0\0\0\0\0\0\0:\1\0\0\0\0\0\0", iov_len=16}, {iov_base="\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\1\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", iov_len=128}], 2) = 144 <0.000038>
327   21:16:26.828845 read(4,  <detached ...>
```

**Additional context**

Thanks for the awesome software!

**Comment by trapexit** [comment]
Touch is does not do a write.

**Comment by broken** [comment]
What technically constitutes a write? I also tried saving a file in vim, fallocate, pipe to file, and copying a file to that directory.

```bash
/mnt/storage/tmp# fallocate -l 1M test.img
fallocate: cannot open test.img: No space left on device
/mnt/storage/tmp# echo 'hi' > test
bash: test: No space left on device
/mnt/storage/tmp# echo 'hi' > /tmp/test; cp /tmp/test .
cp: cannot create regular file './test': No space left on device
```

**Comment by trapexit** [comment]
A write is a write. The literal call to write. 

> defaults,moveonenospc=true,minfreespace=50G,fsname=mergerfs

You are using epmfs policy. Which means it will ignore all branches that don't have the path in question.  And you have minfreespace set. As the docs say if all branches are filtered you'll get an error like ENOSPC or ENOENT or EROSYS depending on which error is found for or maybe last. There isn't a great way to choose.

You have minfreespace set to 50G. Only disk7 has more than 50G free. Have you confirmed, because you are using epmfs, that the path you are writing to exists on disk7?

**Comment by broken** [comment]
I understand that. I thought the point of `moveonenospc` was that if I get an ENOSPC error, it will change the policy.

From the documentation:
> moveonenospc=BOOL|POLICY: When enabled if a write fails with ENOSPC (no space left on device) or EDQUOT (disk quota exceeded) the policy selected will run to find a new location for the file.

According to this, setting `moveonenospc=true` or `moveonenospc=mfs` one would expect that the file would be created despite the directory not existing, because the write failed on the drive due to lack of space, so according to the documentation, the epmfs policy will be replaced by mfs and the file should be written according to the new policy. As I stated in the description, I already verified the file gets written as expected when I set `category.create=mfs`.

**Comment by trapexit** [comment]
> I understand that. I thought the point of moveonenospc was that if I get an ENOSPC error, it will change the policy.

No. That's not what it says. You are ignoring the fact that it says "if a write fails."

> According to this, setting moveonenospc=true or moveonenospc=mfs one would expect that the file would be created despite the directory not existing,

Why would you expect that? That's not what the docs say. You are not writing. You haven't gotten that far. Writing a file requires multiple steps. Creating the file is the first one. mergerfs can't tell the future. It doesn't know what some piece of software is going to do down the road. You told mergerfs via using create policy epmfs "if a create is requested only consider the branches with the basepath of the file to be created" and that is what is happening. If that isn't the behavior you want then don't use that policy. The docs explain exactly what the features do and do them as described. If you want a different behavior use different options. This is not a bug.

**Comment by broken** [comment]
> Why would you expect that?

Because I expected creating the file is part of the write process. I see now that pedantically these are separate. That's all you needed to say.

**Comment by trapexit** [comment]
It is not "pedantic" to maintain different features for different situations and usecases. Just because you don't think you need to worry about them or care about such situations doesn't mean other users don't. Doing what you think it should do would fundamentally break minfreespace and ep policies. Both of which are entirely in your control. There exists numerous other options  for you to get what you want. Most obviously msp policies.

**Comment by broken** [comment]
> Just because you don't think you need to worry about them or care about such situations doesn't mean other users don't.

When did I remotely say this? I don't believe this. I think it's great there are so many options. I never once said you should change anything. In fact, if you read my description, my two theories for the issue were not even problems with the library.

I erroneously believed creating a file could be thought of as an initial step in the write process. I even said:

> Despite reading the documentation, I am not understanding the option [moveonenospc] correctly.

Which I didn't. All you needed to say was, "You are not understanding correctly, creating a file and writing to a file are treated distinctly". Instead, all of your responses have been belligerent and condescending.

**Comment by trapexit** [comment]
> When did I remotely say this? 

1) You opened a bug report. You didn't start a discussion saying you were unsure of the behavior. You didn't mark the issue ticket as a question. That suggests pretty strongly you believe the behavior is incorrect or the documentation is wrong.
2) "I would expect that the test file and paths are created on the drive with space."
3) You call the behaviors pedantic, arguing about the behavior, and saying you expect it to work contrary to the docs. My first response was clarifying what the feature did and quoting the docs.
4) "According to this, setting moveonenospc=true or moveonenospc=mfs one would expect that the file would be created despite the directory not existing,"
5)  "I see now that pedantically these are separate."

> Which I didn't.

And I pointed out the correct interpretation in my first response and there is extensive documentation on this very situation. I'm not omniscient. I can't know you don't know how filesystems work or that you didn't read the whole docs or whatever.

>belligerent and condescending

WTF is belligerent and condescending about answering question with exact quotes from the docs and more? And asking why you believe it should work contrary to the documentation? How am I supposed to know why you thought that and possibly improve the docs unless I ask? I went out of my way on my day off to respond including quoting you the docs which have all the details about how these things work and moreso. If you think that's belligerent and condescending so be it. 

You were the one who said " I see now that pedantically these are separate. That's all you needed to say." which is 1) wrong. they are separate because the must be to do what they advertise to do. 2) rude. I don't need that. This is my hobby. And if people aren't going to respect me and my time then perhaps I don't need to continue it in this form. You might think this an over reaction but this kind of thing happens all the time and it is getting extremely tiring.
